runtime/src/gpu/rocm/gpu-rocm.c

```diff
@@ -27,67 +27,497 @@
+                              blk_dim_x, blk_dim_y, blk_dim_z,
+                              nargs, args);
 }
-void chpl_gpu_impl_mem_free(void* memAlloc) {}
+
+inline void chpl_gpu_impl_launch_kernel_flat(int ln, int32_t fn,
+                                             const char* name,
+                                             int num_threads,
+                                             int blk_dim,
+                                             int nargs,
+                                             va_list args) {
+  int grd_dim = (num_threads+blk_dim-1)/blk_dim;
+
+  chpl_gpu_launch_kernel_help(ln, fn,
+                              name,
+                              grd_dim, 1, 1,
+                              blk_dim, 1, 1,
+                              nargs, args);
+}
+
 void* chpl_gpu_impl_memmove(void* dst, const void* src, size_t n) {
+  #ifdef CHPL_GPU_MEM_STRATEGY_ARRAY_ON_DEVICE
+  bool dst_on_host = chpl_gpu_impl_is_host_ptr(dst);
+  bool src_on_host = chpl_gpu_impl_is_host_ptr(src);
+
+  if (!dst_on_host && !src_on_host) {
+    chpl_gpu_impl_copy_device_to_device(dst, src, n);
+    return dst;
+  }
+  else if (!dst_on_host) {
+    chpl_gpu_impl_copy_host_to_device(dst, src, n);
+    return dst;
+  }
+  else if (!src_on_host) {
+    chpl_gpu_impl_copy_device_to_host(dst, src, n);
+    return dst;
+  }
+  else {
+    assert(dst_on_host && src_on_host);
+    return memmove(dst, src, n);
+  }
+  #else
+
+  // for unified memory strategy we don't want to generate calls to copy
+  // data from the device to host (since it can just be accessed directly)
   return memmove(dst, src, n);
+  #endif
 }
 
-void chpl_gpu_impl_hostmem_register(void *memAlloc, size_t size) { }
+void chpl_gpu_impl_copy_device_to_host(void* dst, const void* src, size_t n) {
+  assert(chpl_gpu_is_device_ptr(src));
 
-void chpl_gpu_impl_copy_device_to_host(void* dst, const void* src, size_t n) {}
-void chpl_gpu_impl_copy_host_to_device(void* dst, const void* src, size_t n) {}
+  CUDA_CALL(hipMemcpyDtoH(dst, (hipDeviceptr_t)src, n));
+}
 
-void chpl_gpu_impl_copy_device_to_device(void* dst, const void* src, size_t n) {}
+void chpl_gpu_impl_copy_host_to_device(void* dst, const void* src, size_t n) {
+  assert(chpl_gpu_is_device_ptr(dst));
 
-bool chpl_gpu_impl_is_device_ptr(const void* ptr) {
-  return false;
+  CUDA_CALL(hipMemcpyHtoD((hipDeviceptr_t)dst, src, n));
 }
 
-bool chpl_gpu_impl_is_host_ptr(const void* ptr) {
-  return false;
+void chpl_gpu_impl_copy_device_to_device(void* dst, const void* src, size_t n) {
+  assert(chpl_gpu_is_device_ptr(dst) && chpl_gpu_is_device_ptr(src));
+
+  CUDA_CALL(hipMemcpyDtoD((hipDeviceptr_t)dst, (hipDeviceptr_t)src, n));
+}
+
+
+void* chpl_gpu_impl_comm_async(void *dst, void *src, size_t n) {
+/*  hipStream_t stream;
+  hipStreamCreateWithFlags(&stream, hipStreamNonBlocking);
+  cuMemcpyAsync((hipDeviceptr_t)dst, (hipDeviceptr_t)src, n, stream);
+  return stream;*/
```

Comment:
is this one of the HIP limitations?

Reply:
Here's the map of CUDA to HIP functions:

https://docs.amd.com/bundle/HIP_Supported_CUDA_API_Guide_v5.2/page/CUDA_Driver_API_functions_supported_by_HIP.html

The entry for `cuMemcpyAsync` is blank. There are `hipMemcpyDtoDAsync`, `hipMemcpyDtoHAsync`, and `hipMemcpyHtoDAsync` functions so we could support it just by adding a little logic to determine what the pointer type is and dispatching appropriately.

Reply:
Ah, that's unfortunate. I think we have a similar logic somewhere in our code that we can reuse. No biggie.